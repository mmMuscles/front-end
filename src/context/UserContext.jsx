import { createContext, useContext, useMemo, useState } from 'react';
import { supabase } from '../services/supabaseClient';

const UserContext = createContext();

const UserProvider = ({ children }) => {
  const session = supabase.auth.session()
  const [profile, setProfile] = useState({})
  const [user, setUser] = useState(
    session ? { id: session.user.id, email: session.user.email } : {}
  );
  
  const value = useMemo(() => ({ user, setUser, profile, setProfile}), [user, profile]);

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

const useUser = () => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }
  return context;
};

export { UserProvider, useUser };
