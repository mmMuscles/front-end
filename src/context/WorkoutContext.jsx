import { useContext, createContext, useState } from "react";



const workoutCtx = createContext();


export const WorkoutProvider = ({ children }) => {
    const [workouts, setWorkouts] = useState([])

return <workoutCtx.Provider value={{workouts, setWorkouts}}>{children}</workoutCtx.Provider>
}



export const useWorkout = () => {
    const context = useContext(workoutCtx)

    if(context === undefined){
        throw new Error('Invalid Provider')
    }

    return context;
}
