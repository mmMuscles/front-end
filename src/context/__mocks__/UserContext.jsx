import { createContext, useContext, useMemo, useState } from "react";

const UserContext = createContext();

const UserProvider = ({ children, mockUser }) => {
  const [profile, setProfile] = useState({});
  const [user, setUser] = useState(
    mockUser ? { id: mockUser.id, email: mockUser.email } : {}
  );

  const value = useMemo(
    () => ({ user, setUser, profile, setProfile }),
    [user, profile]
  );

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

const useUser = () => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
};

export { UserProvider, useUser };
