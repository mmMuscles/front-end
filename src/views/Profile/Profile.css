.profile-back {
    background-image: url('../../assets/download.jpeg');
    background-size: cover;
}

.profile-back label {
    margin: 1rem;
    color: white;
}

/* * {
    border: violet 1px solid;
} */

/* .profile-form {
    display: flex;
    flex-direction: column;
} */