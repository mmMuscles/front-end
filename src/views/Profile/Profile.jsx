import { useState, useEffect } from "react";
import { supabase } from "../../services/supabaseClient";

export default function Profile({ session }) {
  const [loading, setLoading] = useState(true);
  const [username, setUsername] = useState(null);
  const [goal_1, setGoal1] = useState(null);
  const [goal_2, setGoal2] = useState(null);
  const [goal_3, setGoal3] = useState(null);
  const [weight, setWeight] = useState(null);

  useEffect(() => {
    getProfile();
  }, [session]);

  async function getProfile() {
    try {
      setLoading(true);
      const user = supabase.auth.user();

      let { data, error, status } = await supabase
        .from("profiles")
        .select(`username, goal_1, goal_2, goal_3, weight`)
        .eq("email", user.email)
        .single();

      if (error && status === 406) {
        let newProfile = await supabase.from("profiles").insert({
          email: user.email,
          user_id: user.id,
        });
        data = newProfile.data;
      } else if (error && status !== 406) {
        throw error;
      }

      if (data) {
        setUsername(data.username);
        setGoal1(data.goal_1);
        setGoal2(data.goal_2);
        setGoal3(data.goal_3);
        setWeight(data.weight);
      }
    } catch (error) {
      alert(error.message);
    } finally {
      setLoading(false);
    }
  }

  async function updateProfile({ username, goal_1, goal_2, goal_3, weight }) {
    try {
      setLoading(true);
      const user = supabase.auth.user();

      const updates = {
        username,
        goal_1,
        goal_2,
        goal_3,
        weight,
        updated_at: new Date(),
      };

      let { error } = await supabase
        .from("profiles")
        .update(updates, {
          returning: "minimal", // Don't return the value after inserting
        })
        .match({
          email: user.email,
          user_id: user.id,
        });

      if (error) {
        throw error;
      }
    } catch (error) {
      alert(error.message);
    } finally {
      setLoading(false);
    }
  }

  return (
    <div className="form-widget">
      
      {/* <div>
        <label htmlFor="email">Email</label>
        <input id="email" type="text" value={session.user.email} disabled />
      </div> */}
      <div>
        <label htmlFor="username">Name: </label>
        <input
          id="username"
          type="text"
          value={username || ""}
          onChange={(e) => setUsername(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="goal_1">Primary Goal: </label>
        <input
          id="goal_1"
          type="goal_1"
          value={goal_1 || ""}
          onChange={(e) => setGoal1(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="goal_2">Secondary Goal: </label>
        <input
          id="goal_2"
          type="goal_2"
          value={goal_2 || ""}
          onChange={(e) => setGoal2(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="goal_3">Other Goals: </label>
        <input
          id="goal_3"
          type="goal_3"
          value={goal_3 || ""}
          onChange={(e) => setGoal3(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="weight">Weight: </label>
        <input
          id="weight"
          type="weight"
          value={weight || ""}
          onChange={(e) => setWeight(e.target.value)}
        />
      </div>

      <div>
        <button
          // className="button block primary"
          onClick={() => updateProfile({ username, goal_1, goal_2, goal_3, weight })}
          disabled={loading}
        >
          {loading ? "Loading ..." : "Update"}
        </button>
      </div>

      {/* <div>
        <button className="button block" onClick={() => supabase.auth.signOut()}>
          Sign Out
        </button>
      </div> */}
    </div>
  );
}
