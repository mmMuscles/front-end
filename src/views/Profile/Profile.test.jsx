import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { UserProvider } from "../../context/UserContext";
import { WorkoutProvider } from "../../context/WorkoutContext";
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import Profile from "./Profile";

let mockResponse = {
  goal_1: "Lift a car",
  goal_2: "beat up my bully",
  username: "Karl",
  weight: "140",
  weight_goal: "260"
}

const server = setupServer(
  // Describe the requests to mock.
  rest.get('https://psdgtyeifanapnczvbzn.supabase.co/rest/v1/profiles', (req, res, ctx) => {
    return res(
      ctx.json(mockResponse))
  }),
)

beforeAll(() => {
  // Establish requests interception layer before all tests.
  server.listen()
})
afterAll(() => {
  // Clean up after all tests are done, preventing this
  // interception layer from affecting irrelevant tests.
  server.close()
})

// Combined behavior & snapshot test
it("allow user to change profile inputs", async () => {
  const { container } = render(
    <UserProvider >
      <WorkoutProvider>
        <Profile />
      </WorkoutProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();

  const nameField = screen.getByLabelText("Name");
  const primaryGoalField = screen.getByLabelText("Primary Goal");
  const secondaryGoalField = screen.getByLabelText("Secondary Goal");
  const weightField = screen.getByLabelText("Weight");
  const weightGoalField = screen.getByLabelText("Weight goal");
  const updateBtn = screen.getByRole("button", { name: "updating" });

  fireEvent.change(nameField, {
    target: { value: "karl new" },
  });

  mockResponse.username = 'karl new'

  fireEvent.change(primaryGoalField, {
    target: { value: "win" },
  });

  mockResponse.goal_1 = 'win'

  fireEvent.change(secondaryGoalField, {
    target: { value: "win more" },
  });

  mockResponse.goal_2 = 'win more'

  fireEvent.change(weightField, {
    target: { value: "130" },
  });

  mockResponse.weight = '130'

  fireEvent.change(weightGoalField, {
    target: { value: "140" },
  });

  mockResponse.weight_goal = '140'

  fireEvent.click(updateBtn);
  await screen.findAllByDisplayValue('140')

  expect(nameField).toHaveValue("karl new");
  expect(primaryGoalField).toHaveValue("win");
  expect(secondaryGoalField).toHaveValue("win more");
  expect(weightField).toHaveValue("130");
  expect(weightGoalField).toHaveValue("140");


  return waitFor(() => {
    screen.getByDisplayValue("karl new");
  });
});
