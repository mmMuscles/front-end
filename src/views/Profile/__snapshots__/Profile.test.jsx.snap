// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allow user to change profile inputs 1`] = `
<div>
  <main
    class="profile-back h-screen flex flex-col"
  >
    <form
      class="profile-form"
    >
      <section>
        <label
          for="username"
        >
          Name 
        </label>
        <input
          id="username"
          required=""
          type="text"
          value=""
        />
      </section>
      <section>
        <label
          for="goal_1"
        >
          Primary Goal 
        </label>
        <input
          id="goal_1"
          type="goal_1"
          value=""
        />
      </section>
      <section>
        <label
          for="goal_2"
        >
          Secondary Goal 
        </label>
        <input
          id="goal_2"
          type="goal_2"
          value=""
        />
      </section>
      <section>
        <label
          for="weight"
        >
          Weight 
        </label>
        <input
          id="weight"
          type="weight"
          value=""
        />
      </section>
      <section>
        <label
          for="weight_goal"
        >
          Weight goal 
        </label>
        <input
          id="weight_goal"
          type="weight_goal"
          value=""
        />
      </section>
      <button
        class="hover:bg-blue-700 text-white font-bold py-1 px-4 ml-96 rounded w-24"
        disabled=""
      >
        updating
      </button>
    </form>
  </main>
</div>
`;
